36_SpringBootでの例外処理

javachead opened the pull request:
**お世話になっております。

講義動画36の課題である、<SpringBootでの例外処理>の実装を行いました。
ご確認よろしくお願いいたします。**

以下GlobalExceptionHandlerの【入力値バリデーションエラー】結果です
![スクリーンショット 2025-01-29 015950](https://github.com/user-attachments/assets/7ff3176f-a826-4b16-956e-56c6a762febb)

javachead committed:
Add Java CI workflow

javachead committed:
36_Student、StudentCourse、および ErrorResponse に新しいバリデーションルールを追加し、プロパティをリファクタリング。Scorecards と CodeQL 分析のための GitHub ワークフローを追加。動的な更新および挿入を改善するために SQL マッピングを改良。@RestControllerAdviceのクラスの実装。

javachead commented:
↑初めて
CodeQL Analysis
Code scanning results
Java CI使いましたが只今状況確認中です💦

javachead committed:
ResponseEntity<?> → Map<String, Object>へ型修正

javachead committed:
Add GitHub Actions workflow

javachead committed:
Analyze code with CodeQL (java)とCodeQL Analysisの再設定

YudaiNoda3576 commented on src/main/java/raisetech/student/service/StudentService.java:
```diff
@@ -1,107 +1,110 @@
+     * 学生情報を保存または更新するプライベートメソッド。
+     * @param student 更新または保存する学生情報
+     */
+    private void saveOrUpdateStudent(Student student) {
```
saveという命名については、`insert または create`の二つの意味を含んでいる場合が多いですかね。
Spring Data JPA などのsaveメソッドもそのような`insert または create` を行いますし。
なので、単に、saveStudentという命名で良いかと思います。

javachead replied:
ありがとうございます。修正しました

YudaiNoda3576 commented on src/main/java/raisetech/student/service/StudentCourseService.java:
```diff
@@ -68,31 +68,18 @@ public List<StudentCourse> updateByStudentId(Integer studentId) {
-                            }
-                    );
-        }
+    private void processCourses(List<StudentCourse> targetCourses, List<StudentCourse> existingCourses,
```
メソッドの命名も変更されている理由ってなんでしょうか？

YudaiNoda3576 replied:
あ、一応理由だけお伺いしたいです！ @javachead さんの考えを聞きたく！
handleCourses よりもprocessCoursesの方が良いと思って修正されていると思うので🙏

javachead replied:
handleCourses という名前は「ハンドルする（対応する）」という抽象的な意味を持ち、メソッドの具体的な動作が明確ではないとGPTに言われたため命名修正しました💦
一方で processCourses のprocessは日本語で「処理する」という意味を持ち、よりメソッドが何をするかを表現しているとのことでした。

YudaiNoda3576 replied:
抽象度という観点であれば、handleもprocessもあまり変わらない印象がありますね、、、
ChatGPTの回答は鵜呑みにせずに自分で本当にその通りか？という観点でレビュー&チェックが必要なのでお気をつけください🙏
で、このメソッドがやっていることは保存か更新なので、ここのメソッドこそ「save」じゃないですかね？

javachead replied:
💦注意いたします。
「検索結果が本当にその通りか？」という観点で、違う媒体での検索と検証にもっと時間をかけたいと思います。

YudaiNoda3576 commented on src/main/java/raisetech/student/service/StudentCourseService.java:
```diff
@@ -68,31 +68,18 @@ public List<StudentCourse> updateByStudentId(Integer studentId) {
-                    );
-        }
+    private void processCourses(List<StudentCourse> targetCourses, List<StudentCourse> existingCourses,
+                                Long studentId, Merger<StudentCourse> merger) {
```
mergerという命名は抽象的すぎる気がします

YudaiNoda3576 commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -121,6 +123,50 @@ public ResponseEntity<StudentResponse> getStudent(@PathVariable @Min(1) @Max(999
+
+        try {
+            // 学生データを取得（存在確認のため）
+            studentService.getStudentById(id);
```
ここの処理は必要ですかね？

YudaiNoda3576 commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -121,6 +123,50 @@ public ResponseEntity<StudentResponse> getStudent(@PathVariable @Min(1) @Max(999
+     * @return 更新結果のレスポンス
+     */
+    @PutMapping("/{id}")
+    public ResponseEntity<?> updateStudent(
```
型をジェネリクスにしているのはどのような理由でしょうか？

YudaiNoda3576 commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -121,6 +123,50 @@ public ResponseEntity<StudentResponse> getStudent(@PathVariable @Min(1) @Max(999
+            Student updatedStudent = studentDetail.getStudent();
+            updatedStudent.setId(id); // 更新する場合もIDは固定
+            studentService.updateStudent(updatedStudent);
+            logger.info("学生データを更新しました: {}", updatedStudent);
```
ここでログを出力しているのはなぜでしょうか？

YudaiNoda3576 replied:
全体的に過剰にログを出力していると感じました🙏(修正は不要ですが、以下のことは頭の片隅に置いてください🙏)
ログはエラーが発生した場合などの調査に使用するものなのですが、不要なログが多いと調査の妨げになったり、インフラのコストが増大したりなどの弊害も生むので現場では必要以上にログは出力しないようにした方が良いです！

YudaiNoda3576 commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -121,6 +123,50 @@ public ResponseEntity<StudentResponse> getStudent(@PathVariable @Min(1) @Max(999
+
+            // コース情報更新処理
+            List<StudentCourse> updatedCourses = studentDetail.getStudentCourses();
+            studentCourseService.saveCourses(updatedCourses, id); // saveCoursesを利用
```
Serviceクラスでは更新と新規登録を一括で行うメソッドが定義されていたと思いますが、それを使用していないのはどのような理由でしょうか？？

javachead replied:
リファクタリングしているうちにおかしくなってしまったんだとおもいます💦
以下修正箇所のコミットハッシュです
②commit [4e9b3053a5cf4119631a8f2d9dbb5cee961f05c2](https://github.com/javachead/ServiceController/pull/12/commits/4e9b3053a5cf4119631a8f2d9dbb5cee961f05c2)

YudaiNoda3576 commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -162,7 +208,7 @@ public ResponseEntity<StudentDeleteResponse> removeStudent(@PathVariable @Min(1)
             logger.info("指定された学生を削除しました。ID: {}", id);
 
-            return ResponseEntity.ok(new StudentDeleteResponse("学生が削除されました"));
+            return ResponseEntity.ok(new StudentDeleteResponse("学生が論理削除されました"));
```
論理削除されたかどうかについてクライアントサイドに明示する必要はないかなと思います！
というのもクライアントから見たら全て削除されているように見えるので、それが物理的か論理的かというところは知る必要がないからですね。
逆に論理削除が採用されていることが第三者に伝わってしまうと意図しないセキュリティホールになる可能性もなくはないので明示しない方が無難ですね！

YudaiNoda3576 commented on src/main/java/raisetech/student/data/Student.java:
```diff
@@ -14,19 +20,40 @@
+
+    private String remark; // 備考は任意
+
+    private boolean isDeleted; // 論理削除フラグのためバリデーション不要
```
論理削除のためバリデーションは不要というのはどういう理屈ですかね？？
例えばtrue/false以外の値が渡ってきた場合にエラーが発生する可能性もあり得そうな気もします！

YudaiNoda3576 reviewed (commented):
コメントしてます〜

javachead commented on src/main/java/raisetech/student/service/StudentService.java:
```diff
@@ -1,107 +1,110 @@
-     * IDが空の場合は新規作成、存在する場合は更新処理を行う。
-     * - 新規作成時は挙動がシンプルで変更点なし。
-     * - 更新時には `updateStudentDetails` を介するため、検証ロジックやロギングが一貫する。
+     * 学生情報およびコース情報を保存または更新するメソッド。
```
ありがとうございます。判断に迷った時は先輩エンジニアさんに相談します

YudaiNoda3576 replied:
メソッドは修正せずにそのままいきますか？
保存と更新はまとめてしまって良いと思いますが、学生情報とコース情報の扱いはServeiceクラスが独立しているので、それぞれのServieceクラスに保存・更新メソッドを定義してそれらを呼び出す形にする必要はあるかなと思います！

javachead committed:
①Student のフィールドに新しいバリデーション注釈を導入。
サービス内のメソッドをリファクタリングし、可読性と再利用性を向上。
②コースの処理フローを変更し、Merger を CourseUpdater に置き換え。
③コントローラのレスポンスを更新
一貫性を持たせるためにレスポンスを更新。
例外処理を簡素化し、処理を統一化。
④データ操作やエラーハンドリングの効率化

javachead commented on src/main/java/raisetech/student/service/StudentService.java:
```diff
@@ -1,107 +1,110 @@
+        existingStudent.setEmail(updatedStudent.getEmail());
+        existingStudent.setAge(updatedStudent.getAge());
+        existingStudent.setKanaName(updatedStudent.getKanaName());
+        existingStudent.setArea(updatedStudent.getArea());
```
勉強になります修正しました(^▽^)/
以下、修正箇所のコミットハッシュです
①commit [61f2f71db326dcdc3924890e6344b69f58b55a54](https://github.com/javachead/ServiceController/pull/12/commits/61f2f71db326dcdc3924890e6344b69f58b55a54)

javachead commented on src/main/java/raisetech/student/service/StudentService.java:
```diff
@@ -1,107 +1,110 @@
+        // 任意フィールドの更新（Optional を活用して null チェックを簡略化）
+        Optional.ofNullable(updatedStudent.getNickname()).ifPresent(existingStudent::setNickname);
+        Optional.ofNullable(updatedStudent.getSex()).ifPresent(existingStudent::setSex);
+        Optional.ofNullable(updatedStudent.getRemark()).ifPresent(existingStudent::setRemark);
```
修正しました(^▽^)/
以下、修正箇所のコミットハッシュです
①commit [61f2f71db326dcdc3924890e6344b69f58b55a54](https://github.com/javachead/ServiceController/pull/12/commits/61f2f71db326dcdc3924890e6344b69f58b55a54)

javachead commented on src/main/java/raisetech/student/service/StudentService.java:
```diff
@@ -1,107 +1,110 @@
      * @param student 保存対象の学生情報
      */
-    public void saveStudent(Student student) {
+    public void saveStudentAndCourses(Student student) {
```
修正しました(^▽^)/
以下、修正箇所のコミットハッシュです
①commit [61f2f71db326dcdc3924890e6344b69f58b55a54](https://github.com/javachead/ServiceController/pull/12/commits/61f2f71db326dcdc3924890e6344b69f58b55a54)

javachead committed:
学生更新ロジックをリファクタリングし、サービス層を通じて一括保存が行えるよう修正。また、エラーハンドリングを強化しました

YudaiNoda3576 commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -0,0 +1,37 @@
+          languages: ${{ matrix.language }}
+
+      - name: Perform CodeQL Analysis
+        uses: github/codeql-action/analyze@v2
```
@javachead 
確認しました！
修正内容としては問題ないのですが、コミットの粒度が大きい印象がありました。
基本的にはこういったコメント:コミット=1:1になるようにするとやり取りがスムーズになることが多いです！
小さい修正でコミットを分けるまでもなかったり、似たような指摘の場合はまとめてコミットしてしまう場合がありますが、、、

javachead committed:
単一責任の原則に沿って修正
サービスメソッドを簡略化し、冗長性を削減して、責任を一元化するために、学生およびコース管理ロジックをリファクタリングしました。

YudaiNoda3576 commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -68,25 +68,16 @@ public ResponseEntity<StudentAddResponse> createStudent(@RequestBody @Valid Stud
-                logger.info("学生コースデータを保存しました: {}", course);
-            });
+            // 学生情報と関連するコース情報を登録
+            studentDetailService.saveStudentAndCourses(studentDetail);
```
saveStudentAndCourses は単一責任になってますかね？？
StudentServieとStudentCourceそれぞれにsaveメソッドが定義されていて、ControllerでTransaction管理下に置いたうえで呼び出されるのが適切かなと思います🙏

javachead replied:
なんどもコメントさせてしまい申し訳ございません💦
修正しました。

YudaiNoda3576 commented on src/main/java/raisetech/student/service/StudentService.java:
```diff
@@ -1,107 +1,77 @@
-                    log.error("学生名が無効です: {}", student.getName());
-                    return new IllegalArgumentException("学生名は必須です");
-                });
+    @Transactional(readOnly = true)
```
やl.55,l.65はなぜTransaction管理しているのでしょうか？

javachead replied:
データベースを変更する操作のため、外部システムとの通信エラーなどで例外が発生した場合でも（不完全な状態）のデータが残ることを防ぎ整合性を保てるので念のためつけておいた方がいいとの判断です。

YudaiNoda3576 replied:
Getのような読み取り処理の場合でもデータベースを変更する処理に該当しますかね？

javachead replied:
調べました。トランザクション管理下においた際の弊害があり、（1）データベースでのロック時間が伸びて他のトランザクションをブロックする可能性（2）リソース消費の増加（3）デッドロックのリスク（4） 予期せぬロールバック発生　が発生するとありました💦　気づくのが遅くなり申し訳ございません💦💦💦

YudaiNoda3576 commented on .github/workflows/scorecard.yml:
```diff
@@ -0,0 +1,43 @@
+          #git push origin main
+
+          #env:
+          #GITHUB_TOKEN: ${{ secrets.PAT }}
```
まだから行が入っていないファイルが散見されるのでご対応お願いします🙏

javachead replied:
IDEの設定で保存時に空行が含まれるように設定しました💦
以前も設定したはずなんですがうまくできていないようでした。
![スクリーンショット 2025-02-03 155314](https://github.com/user-attachments/assets/d10161b9-f35b-4e74-aa26-01731c232772)

YudaiNoda3576 replied:
一度PRのレビュー依頼する前にセルフレビューするというプロセスを習慣化すると良いですね！
以前指摘した内容がリレビューした時に治っていないパターンは、現場でやると結構嫌がられるパターンなのでご注意を、、、不必要に自分の評価を下げてしまいかねないので、、、

javachead replied:
「以前、指摘した内容がリレビューした時に治っていないパターンは、現場ではかなり嫌がられるパターンなのでご注意を」
←反省しました。このようなご指摘をいただき、大変申し訳ないのと同時にとてもためになるご助言をいただき感謝しております。学校ではこういうことも学びたいと考えております。今回は社会人として欠けた対応だったと思います..💦今後はセルフプレビューの内容も含めてレビュー依頼させていただきます。今後も宜しければ同じようにご教示頂けますと幸いです。

YudaiNoda3576 reviewed (commented):
コメントしております〜

javachead committed:
空行挿入
IDEの設定で保存時に空行が含まれるように設定

javachead committed:
saveStudentAndCoursesを単一責任の原則の基づき分離
トランザクション管理下@Transactionalへ修正

javachead committed:
yml修正

javachead committed:
yml修正2

YudaiNoda3576 commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -58,121 +71,155 @@ public StudentController(StudentCourseService studentCourseService,
+     * @return 削除成功時にメッセージを返す
+     * @throws StudentNotFoundException 学生IDが存在しない場合
      */
+    @Transactional
```
ここは単一テーブルの削除処理なので、呼び出し側よりも定義元のメソッドに@transactionalを付与した方が良いと思います！(呼び出しがわでアノテーションを付与するようになると、いろんなところでアノテーションを定義しないと行けなくなりますし、つけ忘れとかも発生するので)

一方で、上記のような複数件のテーブルに登録・更新・削除を行うような場合は今回のアプローチだと、呼び出し側でTransaction管理しないと不整合が発生することがあるので、その場合は呼び出し側でTransactionalアノテーションを付与する必要があります。

javachead replied:
ありがとうございます。 @Transactionalへの理解が足りませんでした。

・単一作業 → 定義元にだけ付ける。つけ忘れ防止「1つの処理なのでシンプル設定！」
・複数作業 → 呼び出し元にだけ付ける。「全部まとめて元に戻せるように設定！」これを基準にして以後、実装します。

YudaiNoda3576 reviewed (commented):
コメントしてます！
あとまだ空行が入っていないファイルが散見されるので対応のほどお願いします！
レビュー依頼前にFilesChandedタブの差分を確認してみてください！
赤い - アイコンが入っているものが空行が含まれていないファイルになります🙏

javachead commented:
> コメントしてます！ あとまだ空行が入っていないファイルが散見されるので対応のほどお願いします！ レビュー依頼前にFilesChandedタブの差分を確認してみてください！ 赤い - アイコンが入っているものが空行が含まれていないファイルになります🙏

→お手数おかけしております。確認しました！（ 赤い - アイコン）は添付写真画面右のバーに表示されている〇で囲った部分のことで宜しいでしょうか。

＜セルフレビュー＞
インテリジェイのコードフォーマット後、diff画面で 赤い横バーは以下画面上で確認できませんでした。
おそらく修正できたかと存じます。
![スクリーンショット 2025-02-04 175152](https://github.com/user-attachments/assets/835c9320-fefd-4af4-8153-b3ba264bcded)

また、多くのファイルのコード終わりに一行空行が入るなどの修正等を目視で確認しました。

**以下修正箇所のコミットハッシュです↓
commit 36f2daeca2ef1d7b3679692cbf92981a759c6f32**

javachead committed:
コードフォーマッターで空行など修正

javachead commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -58,121 +71,155 @@ public StudentController(StudentCourseService studentCourseService,
+     */
+    @Transactional
+    @PostMapping("/{studentId}/courses")
+    public ResponseEntity<StudentAddResponse> createStudentCourse(
```
Nodaさんのご指摘の通り、別作業が増えることになり手間と不整合が発生すると思います☆作った時にはそう思わなくて...💦
ご教示ありがとうございます。
@PostMapping("/{studentId}/courses")は削除しました。

createStudentの中でStudentService.saveとStudentCourseService.saveが呼ばれ登録する実装に変更しました。

javachead commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -58,121 +71,155 @@ public StudentController(StudentCourseService studentCourseService,
-                studentCourseService.saveCourse(course);
-                logger.info("学生コースデータを保存しました: {}", course);
-            });
+            // 学生IDに基づきコース情報を取得
```
実装しました。

javachead committed:
コードフォーマッターで空行など修正

javachead committed:
①@Transactional位置修正② createStudentの実装修正③@PostMapping("/{studentId}/courses")削除

javachead commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -58,121 +71,155 @@ public StudentController(StudentCourseService studentCourseService,
-            // 学生情報を登録
-            Student student = studentDetail.getStudent();
+            // 学生情報を保存
             studentService.saveStudent(student);
```
**以下修正箇所のコミットハッシュです↓
commit 937704acdf1c66eee03e94c6ec5e765dc490a2e7**
ありがとうございます。命名難しい..

javachead commented on src/main/java/raisetech/student/service/StudentCourseService.java:
```diff
@@ -83,11 +84,12 @@ private void removeUnusedCourses(List<StudentCourse> newCourses, List<StudentCou
      */
-    public void saveCourses(List<StudentCourse> courses, Long studentId) {
+    public void saveCourse(List<StudentCourse> courses, Long studentId) {
         log.info("学生ID {} に紐づくコース情報を保存または更新します。件数: {}", studentId, courses.size());
```
saveCourses→saveCourseへ変更しました。今のところコースが１つの実装だった為、ほかのところのSもとるべきだったと思います。添削しずらい要素を作ってしまいました..申し訳ございません💦

javachead commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -71,51 +71,42 @@ public StudentController(StudentCourseService studentCourseService,
+                Long studentId = student.getId(); // StudentオブジェクトからIDを取得
+                saveCourse.forEach(course -> course.setStudentId(studentId)); // 各コースにstudentIdを設定
+                studentCourseService.saveCourse(saveCourse, studentId); // studentIdを渡す
+            }
```
＜セルフレビュー＞
コースの保存処理実装しました。

javachead commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -71,51 +71,42 @@ public StudentController(StudentCourseService studentCourseService,
-            return ResponseEntity.status(HttpStatus.CREATED)
-                    .body(new StudentAddResponse("新しいコースが登録されました", null));
-        } catch (Exception e) {
-            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
```
＜セルフレビュー＞
@PostMapping("/{studentId}/courses")は削除しました。

javachead commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -71,51 +71,42 @@ public StudentController(StudentCourseService studentCourseService,
+            // 2. 学生に紐づくコース情報を保存
+            List<StudentCourse> saveCourse = student.getStudentCourses(); // コース情報を取得
+            if (saveCourse != null && !saveCourse.isEmpty()) {
+                Long studentId = student.getId(); // StudentオブジェクトからIDを取得
```
＜セルフレビュー＞
コースの保存処理実装しました。

javachead commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -71,51 +71,42 @@ public StudentController(StudentCourseService studentCourseService,
-    @Transactional
-    @PostMapping("/{studentId}/courses")
-    public ResponseEntity<StudentAddResponse> createStudentCourse(
-            @PathVariable Long studentId,
```
＜セルフレビュー＞
@PostMapping("/{studentId}/courses")は削除しました。

YudaiNoda3576 reviewed (commented):
コメントしております！

javachead commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -58,121 +71,144 @@ public StudentController(StudentCourseService studentCourseService,
+            if (saveCourse != null && !saveCourse.isEmpty()) {
+                Long studentId = student.getId(); // StudentオブジェクトからIDを取得
+                saveCourse.forEach(course -> course.setStudentId(studentId)); // 各コースにstudentIdを設定
+                studentCourseService.saveCourse(saveCourse, studentId); // studentIdを渡す
```
もともとsaveCourseとsaveStudentがあり、saveStudentのほうを前回saveに変更しました。
こちらは修正すると重複メソッドになるため、こちらはそのままにしております。

YudaiNoda3576 replied:
ここは変数名でしたね！
変数にはどのようなデータが格納されているかを表現するのが良いため、saveCourseではなく、courseList みたいな「名詞」で表現した方がわかりやすいかもです！

あと、markdownのリンクの末尾に`)`がないためリンクがそのまま表示されてしまっています。
次回からはこのあたりも正常な表示がされていない場合は修正するように意識してみてください〜
タイポで動作しなくなるのがプログラムなので、こういう細かいところが気になるかならないかがエンジニアとしての素養があるかないかで企業から見られることもありますので🙏

javachead replied:
ご確認ありがとうございます。
**saveCourse　→　courseList へ修正しました。**

あっ💦「タイポで動作しなくなるのがプログラムなので、こういう細かいところが気になるか、ならないかがエンジニアとしての素養があるかないかで企業から見られることもある」←判断要素になるんですね💦このようなお話をたくさん聞けてとても勉強になりました。ありがとうございました。

javachead commented on src/main/java/raisetech/student/service/StudentCourseService.java:
```diff
@@ -21,99 +23,59 @@ public StudentCourseService(StudentCourseRepository studentCourseRepository) {
      * 新規登録・既存更新のコース処理
      * このメソッドはリストで渡されたコースの中で、新しいものを登録し、既存のものを更新します。
-     * @param newCourses 新規または更新対象のコースリスト
+     * コントローラー側の呼び出し側で @Transactionalは実装済み。
```
あっ！変な文化が継承されてしますね..💦勉強になります。

javachead committed:
①JavaDocコメントの修正

javachead committed:
①不要アノテーションの削除

javachead committed:
①不要アノテーションの削除

javachead commented on src/main/java/raisetech/student/dto/ErrorResponse.java:
```diff
@@ -1,6 +1,23 @@
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
```
実装不要でした
ご確認ありがとうございます。削除しました。

修正しました以下修正箇所のコミットハッシュです↓
commit fdb1cd989467a366707779987c329028318be190

javachead commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -58,121 +71,144 @@ public StudentController(StudentCourseService studentCourseService,
+            // 学生データの更新処理
+            Student updatedStudent = studentDetail.getStudent();
+            updatedStudent.setId(id);
+            studentService.updateStudent(updatedStudent);
```
ご教示ありがとうございます。updateメソッドを削除し、saveへ登録＆更新機能をまとめました。

javachead replied:
ご教示ありがとうございます。updateメソッドを削除し、saveにまとめました。

javachead committed:
①updateStudentをsaveメソッドに統合②handleCoursesをsaveCourseメソッドに統合

javachead commented on src/main/java/raisetech/student/service/StudentCourseService.java:
```diff
@@ -30,7 +30,6 @@ public List<StudentCourse> findByStudentId(Long studentId) {
     /**
      * 新規登録・既存更新のコース処理
      * このメソッドはリストで渡されたコースの中で、新しいものを登録し、既存のものを更新します。
-     * コントローラー側の呼び出し側で @Transactionalは実装済み。
```
＜セルフレビュー＞
コメント削除しました

javachead commented on src/main/java/raisetech/student/service/StudentCourseService.java:
```diff
@@ -84,7 +83,6 @@ public void removeUnusedCourses(List<StudentCourse> newCourses, List<StudentCour
      * - 新規コースは登録し、既存のコースは更新。
      * - 渡されたリストに存在しないコースは削除される。
      * - 入力リストが学生IDと整合性がとれている必要あり。
-     * 　コントローラー側の呼び出し側で @Transactionalは実装済みです。
```
＜セルフレビュー＞
コメント削除しました

javachead commented on src/main/java/raisetech/student/service/StudentService.java:
```diff
@@ -30,7 +30,6 @@ public Student getStudentById(Long id) {
 
     /**
      * 学生情報を更新するメソッド（デフォルトのトランザクション）。
-     * コントローラー側の呼び出し側で @Transactionalは実装済み。
```
＜セルフレビュー＞
コメント削除しました

javachead commented on src/main/java/raisetech/student/service/StudentService.java:
```diff
@@ -56,7 +55,6 @@ public void deleteStudentById(Long studentId) {
 
     /**
      * 新しい学生情報を登録します。
-     * コントローラー側の呼び出し側で @Transactionalは実装済みです。
```
＜セルフレビュー＞
コメント削除しました

javachead commented on src/main/java/raisetech/student/dto/ErrorResponse.java:
```diff
@@ -1,16 +1,10 @@
-@Setter
-@NoArgsConstructor
-@AllArgsConstructor
 public class ErrorResponse {
```
＜セルフレビュー＞
アノテーションを削除しました。

削除理由

- @Setter:
フィールドがfinalであり、setメソッドを使用して値変更を行う設計ではないため不要。

- @NoArgsConstructor:
 finalフィールドが存在している場合、引数なしのデフォルトコンストラクタは生成できず、設計上も不要。

- @AllArgsConstructor:
クラスにカスタムコンストラクタを定義しており、そのカスタムロジックをサポートするため自動生成された全フィールドを引数に持つコンストラクタは不要。

javachead committed:
①saveCourse→courseList へ修正しました。

javachead committed:
CodeQL＆Java CI / CodeQLの修正

javachead committed:
CodeQL＆Java CI / CodeQLの修正②
